### PR TITLE
input: add a gpio qdec input driver

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Hi, wrote this for GPIO based rotary encoded (scroller wheels) to feed data to input devices. Originally wrote this in 2021 as a sensor driver but I think it makes sense to have it as an input device as it's probably a common use case (lvgl has a a concept of input encoder as well, ZMK has a driver as well), so I rewrote it as an input driver and added an timeout and idle mode so it should play well with battery powerd devices.

Let me know what you think.

---

Add a GPIO based quadrature decoder driver that reports relative axes movements using the input subsystem.

### CODE REVIEW

File: drivers/input/input_gpio_qdec.c [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+	if (gpio_pin_get_dt(&cfg->gpio[0])) {
+		step |= 0x01;
+	}
+	if (gpio_pin_get_dt(&cfg->gpio[1])) {
+		step |= 0x02;
+	}
+
+	if (data->prev_step == step)
+		return;
```
Comment (gmarull) [inline drivers/input/input_gpio_qdec.c:67]:
```suggestion
	if (data->prev_step == step) {
		return;
	}
```

File: drivers/input/input_gpio_qdec.c [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+	switch ((data->prev_step << 4) | step) {
```
Comment (gmarull) [inline drivers/input/input_gpio_qdec.c:69]:
```suggestion
	switch ((data->prev_step << 4U) | step) {
```

File: drivers/input/input_gpio_qdec.c [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+static void gpio_qdec_irq_setup(const struct device *dev, bool enable)
+{
+	const struct gpio_qdec_config *cfg = dev->config;
+	unsigned int flags = enable ? GPIO_INT_EDGE_BOTH : GPIO_INT_DISABLE;
+	int ret;
+
+	for (int i = 0; i < 2; i++) {
+		const struct gpio_dt_spec *gpio = &cfg->gpio[i];
```
Comment (gmarull) [inline drivers/input/input_gpio_qdec.c:125]: fixed size here (2) sounds like a bad idea

File: drivers/input/input_gpio_qdec.c [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+static int gpio_qdec_init(const struct device *dev)
+{
+	const struct gpio_qdec_config *cfg = dev->config;
+	struct gpio_qdec_data *data = dev->data;
+	int ret;
+
+	data->dev = dev;
+
+	gpio_pin_configure_dt(&cfg->gpio[0], GPIO_INPUT);
+	gpio_pin_configure_dt(&cfg->gpio[1], GPIO_INPUT);
```
Comment (gmarull) [inline drivers/input/input_gpio_qdec.c:173]: error handling missing

File: drivers/input/input_gpio_qdec.c [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+		.steps_per_period = DT_INST_PROP(n, steps_per_period),	\
+		.axis = DT_INST_PROP(n, zephyr_axis),			\
+	};								\
+									\
+	static struct gpio_qdec_data gpio_qdec_data_##n;		\
+									\
+	DEVICE_DT_INST_DEFINE(n, gpio_qdec_init, NULL,			\
+			      &gpio_qdec_data_##n,			\
+			      &gpio_qdec_cfg_##n,			\
+			      POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,	\
```
Comment (gmarull) [inline drivers/input/input_gpio_qdec.c:222]: `SENSOR`?

File: dts/bindings/input/zephyr,gpio-qdec.yaml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          compatible = "zephyr,gpio-qdec";
+          gpios = <&gpio0 14 (GPIO_PULL_UP | GPIO_ACTIVE_HIGH)>,
+                  <&gpio0 13 (GPIO_PULL_UP | GPIO_ACTIVE_HIGH)>;
+          steps-per-period = <4>;
+          zephyr,axis = <INPUT_REL_Y>;
+          sample-time-us = <2000>;
+          idle-timeout-ms = <200>;
+  };
+
+compatible: "zephyr,gpio-qdec"
```
Comment (gmarull) [inline dts/bindings/input/zephyr,gpio-qdec.yaml:27]: I wonder if the `zephyr,` prefix is really needed, this is a generic peripheral, like `gpio-keys`.

File: drivers/input/input_gpio_qdec.c (hunk 1 above)
Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c]: Sorted.

File: drivers/input/input_gpio_qdec.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c]: Yeah good point, added a `#define GPIO_QDEC_GPIO_NUM 2` and a build check to ensure we have two and only two gpios here.

File: drivers/input/input_gpio_qdec.c (hunk 5 above)
Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c]: Leftover from this code previous life. :-) Fixed, thanks for spotting it.

File: dts/bindings/input/zephyr,gpio-qdec.yaml (hunk 6 above)
Comment (fabiobaltieri) [reply dts/bindings/input/zephyr,gpio-qdec.yaml:27]: I really have no preference but for `gpio-keys` we have both `gpio-keys` for the binding alone and `zephyr,gpio-keys` for the binding associated with the driver. Not too sure why we went that way, on regulator we have `regulator-fixed` but then there's plenty other `zephyr,...` bindings defined. Not sure what to do here.

File: drivers/input/input_gpio_qdec.c [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+	QDEC_LH_HH = 0x13,
+	QDEC_HH_HL = 0x32,
+	QDEC_HL_LL = 0x20,
+
+	/* Negative transitions */
+	QDEC_LL_HL = 0x02,
+	QDEC_LH_LL = 0x10,
+	QDEC_HH_LH = 0x31,
+	QDEC_HL_HH = 0x23,
+};
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c]: Sure why not. Changed to defines.

File: drivers/input/input_gpio_qdec.c [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_gpio_qdec
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/input/input.h>
+#include <zephyr/kernel.h>
+#include <stdlib.h>
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c]: Done, left `stdint.h` in there, it's needed for `abs`.

File: drivers/input/input_gpio_qdec.c (hunk 4 above)
Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c]: Actually there's already a (proper) `gpio_pin_configure_dt` in the cycle below so this is redundant, removed.

File: drivers/input/input_gpio_qdec.c (hunk 2 above)
Comment (fabiobaltieri) [reply drivers/input/input_gpio_qdec.c]: Done.

**Comment by fabiobaltieri** [comment]
All done, one open question on the naming, also added a line for initializing the state at init time based on the actual gpio values, noticed it was missing the first step on an encoder with two steps per period, that fixed it.

File: dts/bindings/input/zephyr,gpio-qdec.yaml (hunk 6 above)
Comment (gmarull) [reply dts/bindings/input/zephyr,gpio-qdec.yaml:27]: @galak ? I think that bindings representing generic hardware, like regulator-fixed, regulator-gpio, gpio-keys, etc do not have any prefix. In the end there's nothing Zephyr specific in there (except input codes?)

Comment (fabiobaltieri) [reply dts/bindings/input/zephyr,gpio-qdec.yaml:27]: @galak ping

Comment (fabiobaltieri) [reply dts/bindings/input/zephyr,gpio-qdec.yaml]: -EGALAKTIMEOUT, dropped the prefix

**Review by fabiobaltieri — COMMENTED** [review]
Dropped the `zephyr,` prefix and rebased.

**Review by gmarull — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]
Looks good!

**Comment by nashif** [comment]
nothing seems to be building it, can you add it to some driver build test?

**Comment by fabiobaltieri** [comment]
> nothing seems to be building it, can you add it to some driver build test?

Yeah good point, I can add an `input` under `build_all` to catch drivers that are not under something else (`i2c`), mind if I do it as a followup?

**Pull request merged**